src/lib/core/decorators/dispatch.ts

```diff
@@ -38,6 +40,39 @@ function dispatchEvent(target: any, key: string | symbol) {
     };
 }
 
+/**
+ * @param event - Wrapped dispatched event
+ * @param dispatch - Function that dispatches event
+ * @param zone - `NgZone` instance
+ */
+function dispatchEventIfZoneIsNotNooped(event: WrappedDispatchedEvent, dispatch: DispatchEventFactory, zone: NgZone): void {
+    const dispatchInsideZone = (event: DispatchedEvent) => zone.run(() => dispatch(event));
+
+    zone.runOutsideAngular(() => {
+        if (isObservable(event)) {
+            event.pipe(first()).subscribe((event) => dispatchInsideZone(event));
```

Comment:
I think you need to combine duplicate code

Reply:
any ideas?

Reply:
duplicate code

![image](https://user-images.githubusercontent.com/12021443/49066108-2d2c5700-f231-11e8-83ac-255d497f2ee6.png)

Reply:
i mean - any ideas how to avoid this?

Reply:
```ts
function dispatchByEvent(event: Event, dispatcher: Function) {
  if (isObservable(event)) {
            event.pipe(first()).subscribe((event) => dispatcher(event));
   } else if (isPromise(event)) {
            event.then((event) => dispatcher(event));
   } else {
            dispatcher(event);
   }
}
```

```ts
dispatchByEvent(event, dispatchInsideZone);
dispatchByEvent(event, dispatchEvent(target, key));
```